src/compton_tools/Compton_Native.cc

```diff
@@ -0,0 +1,625 @@
+  Require(x <= xR);
+
+  // Spacing of interval
+  const T_FP dx = xR - xL;
+  // Left/right basis functions
+  const T_FP bL = (xR - x) / dx;
+  const T_FP bR = (x - xL) / dx;
+
+  // Hermite functions
+  std::array<T_FP, 4> hermite_eval{bL * bL * (3 - 2 * bL), bR * bR * (3 - 2 * bR),
+                                   -dx * bL * bL * (bL - 1), dx * bR * bR * (bR - 1)};
+  return hermite_eval;
+}
+
+//------------------------------------------------------------------------------------------------//
+// SETUP FUNCTIONS
+//------------------------------------------------------------------------------------------------//
+
+//------------------------------------------------------------------------------------------------//
+/*!
+ * \brief Constructor for Compton_Native
+ *
+ * \param[in] filename Name of the binary CSK data file to be read
+ *
+ * Rank 0 reads the binary CSK data file, which fills in the class' data members.  The data members
+ * are then broadcast to other MPI ranks to finish their construction.
+ */
+Compton_Native::Compton_Native(const std::string &filename) {
+  Require(filename.length() > 0U);
+  int rank = rtt_c4::node();
+  constexpr int bcast_rank = 0;
+  int errcode = 0;
+
+  if (rank == bcast_rank) {
+    errcode = read_binary(filename);
+  }
+  broadcast_MPI(errcode);
+
+  Ensure(check_class_invariants());
+}
+
+//------------------------------------------------------------------------------------------------//
+// BINARY READING AND BROADCASTING FUNCTIONS
+//------------------------------------------------------------------------------------------------//
+
+//------------------------------------------------------------------------------------------------//
+/*!
+ * \brief Helper member function that broadcasts CSK data from rank 0 to all ranks and sets data
+ *
+ * \param[in] errcode If non-zero, everyone aborts.
+ *
+ * Uses rtt_c4's broadcast to send arrays and vectors to all ranks with the sparse CSK data,
+ * temperature / energy grids, and sizes
+ */
+void Compton_Native::broadcast_MPI(int errcode) {
+  rtt_c4::global_max(errcode);
+  if (errcode == 1) {
+    // Have all ranks throw a bad file exception
+    std::ifstream f("");
+    f.exceptions(f.failbit);
+  }
+  Insist(errcode == 0, "Non-zero errorcode. Exiting.");
+
+  int rank = rtt_c4::node();
+  constexpr size_t bcast_rank = 0;
+
+  // Broadcast sizes
+  size_t data_size = data_.size();
+  std::array<size_t, 6> pack = {num_temperatures_, num_groups_, num_leg_moments_,
+                                num_evals_,        num_points_, data_size};
+  rtt_c4::broadcast(&pack[0], pack.size(), bcast_rank);
+  size_t p = 0;
+  num_temperatures_ = pack[p++];
+  num_groups_ = pack[p++];
+  num_leg_moments_ = pack[p++];
+  num_evals_ = pack[p++];
+  num_points_ = pack[p++];
+  data_size = pack[p++];
+  Insist(data_size < (1U << 31U), "Unrepresentable with int needed in the broadcast");
+
+  // Derived sizes
+  const auto tsz = static_cast<int>(num_temperatures_);
```

Comment:
Now that we have the new broadcast, I don't think I need this cast to int?

Reply:
Fixed in 6404a2d.

Reply:
I removed this block of `ints` and switched to _iterator-ranged-based_ `broadcast`.  Let me know what you think of this.